server/src/main/scala/org/http4s/server/middleware/CSRFMiddleware.scala

```diff
@@ -0,0 +1,196 @@
+  def withNewToken: Middleware[Request, MaybeResponse, Request, MaybeResponse] =
+    _.andThen(Kleisli(embedNew))
+
+}
+
+object CSRFMiddleware {
+
+  /** An instance of SecureRandom to generate
+    * tokens, properly seeded:
+    * https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/
+    */
+  private lazy val CachedRandom = {
+    val r = new SecureRandom()
+    r.nextBytes(new Array[Byte](20))
+    r
+  }
+
+  val SigningAlgo = "HmacSHA1"
+  val SHA1ByteLen = 20
+  val CSRFTokenLength = 32
+
+  /** Hex encoding digits. Adapted from apache commons Hex.encodeHex **/
+  private val Digits: Array[Char] = Array('0', '1', '2', '3', '4', '5', '6',
+    '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+  protected def find[A](nel: NonEmptyList[A], p: A => Boolean): Option[A] =
+    if (p(nel.head)) {
+      Some(nel.head)
+    } else {
+      nel.tail.find(p)
+    }
+
+  protected def cookieFromHeaders(request: Request,
+    headerName: String): OptionT[Task, Cookie] =
+    OptionT(
+      Task.point(HCookie
+        .from(request.headers)
+        .flatMap(v => CSRFMiddleware.find[Cookie](v.values, _.name == headerName))))
+
+  /** A Constant-time string equality **/
+  def isEqual(s1: String, s2: String): Boolean =
+    MessageDigest.isEqual(s1.getBytes(StandardCharsets.UTF_8),
+      s2.getBytes(StandardCharsets.UTF_8))
+
+  /** Encode a string to a Hexadecimal string representation
+    * Adapted from apache commons Hex.encodeHex
+    */
+  protected def encodeHex(data: Array[Byte]): Array[Char] = {
+    val l = data.length
+    val out = new Array[Char](l << 1)
+    // two characters form the hex value.
+    var i = 0
+    var j = 0
+    while (i < l) {
+      out(j) = Digits((0xF0 & data(i)) >>> 4)
+      j += 1
+      out(j) = Digits(0x0F & data(i))
+      j += 1
+      i += 1
+    }
+    out
+  }
+
+  /** Generate an unsigned CSRF token from a `SecureRandom` **/
+  protected def genTokenString: String = {
+    val bytes = new Array[Byte](CSRFTokenLength)
+    CachedRandom.nextBytes(bytes)
+    new String(encodeHex(bytes))
+  }
+
+  /** Generate a signing Key for the CSRF token **/
+  def generateSigningKey(): SecretKey =
+    KeyGenerator.getInstance(SigningAlgo).generateKey()
+
+  /** Build a new HMACSHA1 Key for our CSRF Middleware
+    * from key bytes. This operation is unsafe, in that
+    * any amount less than 20 bytes will throw an exception when loaded
+    * into `Mac`, and any value above will be truncated (not good for entropy).
+    *
+    * Use for loading a key from a config file, after having generated
+    * one safely
+    *
+    */
+  def buildSigningKey(array: Array[Byte]): Task[SecretKey] =
+    Task.delay(new SecretKeySpec(array.slice(0, SHA1ByteLen), SigningAlgo))
+
+  def unsafeBuildSigningKey(array: Array[Byte]): SecretKey =
+    new SecretKeySpec(array, SigningAlgo)
+
+}
+
+/** A basic CSRF configuration class **/
+case class CSRFConfiguration(
```

Comment:
I've become wary of case classes for configuration, because they don't evolve in a binary compatible way.  It sucks, but I might expand this out into something that looks like what [contraband generates](http://www.scala-sbt.org/contraband/code-generation.html)